commown_devices/models/project_task.py

```diff
@@ -145,18 +145,24 @@ def onchange_lot_id(self):
 
     @api.constrains("stage_id")
     def onchange_stage_id_prevent_contract_resiliation_with_device(self):
-        if self.stage_id and self.contract_id and self.contract_id.quant_nb > 0:
-            check_stage_ids = tuple(
-                self.env.ref(ref).id for ref in CHECK_CONTRACT_QUANT_NB_STAGE_XML_IDS
-            )
-            if self.stage_id.id in check_stage_ids:
-                raise Warning(
-                    _(
-                        "This task can not be moved forward. There are still %s device(s) "
-                        "associated with the contract"
-                    )
-                    % self.contract_id.quant_nb
+        for record in self:
+            if (
+                record.stage_id
+                and record.contract_id
+                and record.contract_id.quant_nb > 0
+            ):
+                check_stage_ids = tuple(
+                    record.env.ref(ref).id
+                    for ref in CHECK_CONTRACT_QUANT_NB_STAGE_XML_IDS
                 )
```

Comment:
Le calcul de check_stage_ids peut sortir de la boucle (en utilisant self.env.ref)

Reply:
On peut implémenter ça de manière pas mal plus performante avec un truc du genre (pseudo-code pas testé) : 
erroneous_tasks = self.search([("id", "in", self.ids), ("contract_id.quant_nb", ">", 0), ("stage_id.id", "in", check_stage_ids)])
if erroneous_tasks
   raise Warning("These tasks can not be moved forward: %s", erroneous_tasks.ids)